Refactor ForeachUtils.h

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52450 wip
* #49039 [wip] Replace optimizers in torch.optim with the ones from torch.optim._multi_tensor
* #52449 Add type promotion to a fast path for foreach APIs
* #52448 Support broadcasting via slow path
* #48224 [wip] Pushed unsupported scenarios to slow path
* #51170 Updated foreach pointwise ops to support complex and bloat16 on CUDA
* #49714 Update foreach max/min
* #51060 Refactor foreach pointwise ops tests to use OpInfo
* #51059 Refactor foreach binary ops tests with tensor lists to use OpInfo
* #51058 Refactor foreach binary ops tests with scalars to use OpInfo
* #49712 Refactor foreach unary ops tests to use OpInfo
* #49250 Update foreach binary ops with a single scalar and list of tensors
* #49249 Update foreach binary ops with a scalar list
* **#51131 Refactor ForeachUtils.h**

--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py

Differential Revision: [D26103904](https://our.internmc.facebook.com/intern/diff/D26103904)

Commit:
Refactor ForeachUtils.h

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUtils.h"




[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py


[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py


[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py


[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -64,155 +64,85 @@ void check_foreach_api_restrictions(TensorList tensors1, TensorList tensors2, Te
-  for (auto t : tensors) {
-    if (!has_same_attributes(expected_device, {t})) {
-      return false;
+bool check_fast_path_restrictions(
+        ArrayRef<TensorList> tensorLists, 
+        ArrayRef<Scalar> scalarList = {}, 
+        bool promote_integer_inputs_to_float = false) {
```
Nit: might be good to document and/or assert preconditions. From what I remember:
- each tensorList in tensorLists must have the same size
- each tensorList in tensorLists must have at least one element
- tensorLists must have at least one element
- scalarList must either be: size 0, size1, or same size as each TensorList in tensorLists

zou3519 replied:
Are you talking about

```
// To go via 'fast' path, several conditions must be satisfied	// To go via 'fast' path, several conditions must be satisfied
// - All tensors must be on the same device	// - All tensors must be on the same device
// - All tensors must have strided layout	// - All tensors must have strided layout
// - All tensors must be non-overlapping and dense	// - All tensors must be non-overlapping and dense
// - Resulting tensor must have the same dtype as the input one	// - Resulting tensor must have the same dtype as the input one
```
?
Because that comment doesn't answer my question: what are the pre-conditions that are true before one calls `check_fast_path_restrictions`?

izdeby replied:
Sorry, my bad
These conditions should be satisfied: 

// Check foreach API restrictions 
// - Tensor lists must be non-empty.
// - All tensors in all lists must have the same dtype.
// - All TensorLists and ScalarLists must have the same number of elements.
// - Corresponding tensors must have the same size.

Its done in `check_foreach_api_restrictions`. 

Added a comment about it.

zou3519 commented on test/test_foreach.py:
```diff
@@ -695,13 +669,14 @@ def test_bool_scalarlist(self, device, dtype):
-                        with self.assertRaisesRegex(RuntimeError, "not implemented for"):
-                            foreach_bin_op(tensors, scalars)
-                        return
+                            with self.assertRaisesRegex(RuntimeError, "Subtraction, the `-` operator, with a bool tensor"):
+                                foreach_bin_op(tensors, scalars)
+                            return
```
Is this indentation intentional?

zou3519 commented on test/test_foreach.py:
```diff
@@ -383,15 +383,6 @@ def test_int_scalarlist(self, device, dtype):
                 scalars = [1 for _ in range(N)]
                 expected = [torch_bin_op(t, s) for t, s in zip(tensors, scalars)]
 
-                # we dont support bool and complex types on CUDA for now
-                if dtype in torch.testing.get_all_complex_dtypes() and self.device_type == 'cuda':
-                    with self.assertRaisesRegex(RuntimeError, "not implemented for"):
-                        foreach_bin_op_(tensors, scalars)
-
-                    with self.assertRaisesRegex(RuntimeError, "not implemented for"):
-                        foreach_bin_op(tensors, scalars)
-                    return
-
```
Why were complex numbers previously not supported? What in this PR makes it so that complex numbers are now supported? Is it because we are delegating complex numbers to the slow path?

zou3519 reviewed (commented):
some questions

Commit:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py

Differential Revision: [D26103904](https://our.internmc.facebook.com/intern/diff/D26103904)

[ghstack-poisoned]

izdeby commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -64,155 +64,85 @@ void check_foreach_api_restrictions(TensorList tensors1, TensorList tensors2, Te
-  for (auto t : tensors) {
-    if (!has_same_attributes(expected_device, {t})) {
-      return false;
+    // We check only tensorLists at index 0 as there no use cases for other indexes yet.
```
Yeah, this is correct. i rewrote it to be more clear

zou3519 replied:
The new thing you wrote now does more work, is that intentional?

izdeby commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -64,155 +64,85 @@ void check_foreach_api_restrictions(TensorList tensors1, TensorList tensors2, Te
-
-    if (t.strides() != expected_strides) {
-      return false;
+bool will_promote_tensor(const Tensor& tensor, Scalar scalar, bool promote_integer_inputs_to_float = false) {
```
Its used in the next PR. moved it there

Commit:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py

Differential Revision: [D26103904](https://our.internmc.facebook.com/intern/diff/D26103904)

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -64,155 +64,69 @@ void check_foreach_api_restrictions(TensorList tensors1, TensorList tensors2, Te
-      return false;
+    for (int j=0; j < tensorLists.size(); j++) {
+      for (int i=0; i < tensorLists[j].size(); i++) {
+        if (scalarList.size() == 1) {
+          if (will_promote_tensor(tensorLists[j][i], scalarList[0])) {
+            return false;
+          }
+        } else if (scalarList.size() > 1) {
+          // Complex scalar list is not supported.
+          if (scalarList[i].isComplex()) {
+            return false;
+          }
```
Is `_foreach_add([torch.tensor(1.+0j), torch.tensor(-1.+0j)], 0+1j)` supposed to go through the fast path? Why does the fast-path support a single complex scalar but not complex scalarList?

izdeby replied:
TensorListMetadata has to be < 4KB - the limit for kernel launch argument and scalar list with complex scalars will not fit in. So for now, we are pushing it via a slow path. Added a comment about this

zou3519 replied:
Got it, thanks for the context! A comment would be great :)

zou3519 commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -64,155 +64,85 @@ void check_foreach_api_restrictions(TensorList tensors1, TensorList tensors2, Te
+               tensor.is_non_overlapping_and_dense();
+    };
+
+    for (const auto& tensorList : tensorLists) {
+      for (const auto& tensor : tensorList) {
```
There's one place that should be fixed:
- check_fast_path_restrictions has an initial 4-space indent

izdeby replied:
Done

Commit:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py

Differential Revision: [D26103904](https://our.internmc.facebook.com/intern/diff/D26103904)

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -64,155 +64,71 @@ void check_foreach_api_restrictions(TensorList tensors1, TensorList tensors2, Te
 
-    if (will_promote_tensor(t, scalar)) {
-      return false;
+    for (int j=0; j < tensorLists.size(); j++) {
+      for (int i=0; i < tensorLists[j].size(); i++) {
+        if (scalarList.size() == 1) {
+          if (will_promote_tensor(tensorLists[j][i], scalarList[0])) {
```
nit: I think how you had the code originally does less work, if it is also correct: https://github.com/pytorch/pytorch/pull/51131#discussion_r565591409 

```
// For all j, tensorList[j][0] have the same shape and dtype. (this was a precondition
// checked by `check_foreach_api_restrictions`). This means we only need to check if
// {tensorList[0][0], tensorList[0][1], tensorList[0][2], ...} do type promotion with scalarLIst.
for (int i=0; i < tensorLists[0].size(); i++) {
  if (scalarList.size() == 1) {
    if (will_promote_tensor(tensorLists[0][i], scalarList[0])) {
      return false;
    }
  } else {
    if (will_promote_tensor(tensorLists[0][i], scalarList[i])) {
      return false;
    }
  }
}
```

Commit:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py

Differential Revision: [D26103904](https://our.internmc.facebook.com/intern/diff/D26103904)

[ghstack-poisoned]

izdeby committed:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py

Differential Revision: [D26103904](https://our.internmc.facebook.com/intern/diff/D26103904)

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py

Differential Revision: [D26103904](https://our.internmc.facebook.com/intern/diff/D26103904)

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py

Differential Revision: [D26103904](https://our.internmc.facebook.com/intern/diff/D26103904)

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py

Differential Revision: [D26103904](https://our.internmc.facebook.com/intern/diff/D26103904)

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py

Differential Revision: [D26103904](https://our.internmc.facebook.com/intern/diff/D26103904)

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py

Differential Revision: [D26103904](https://our.internmc.facebook.com/intern/diff/D26103904)

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py

Differential Revision: [D26103904](https://our.internmc.facebook.com/intern/diff/D26103904)

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py

Differential Revision: [D26103904](https://our.internmc.facebook.com/intern/diff/D26103904)

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py

Differential Revision: [D26103904](https://our.internmc.facebook.com/intern/diff/D26103904)

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py

Differential Revision: [D26103904](https://our.internmc.facebook.com/intern/diff/D26103904)

[ghstack-poisoned]

izdeby committed:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py

Differential Revision: [D26103904](https://our.internmc.facebook.com/intern/diff/D26103904)

[ghstack-poisoned]

izdeby committed:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py

Differential Revision: [D26103904](https://our.internmc.facebook.com/intern/diff/D26103904)

[ghstack-poisoned]

izdeby committed:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py

Differential Revision: [D26103904](https://our.internmc.facebook.com/intern/diff/D26103904)

[ghstack-poisoned]

izdeby committed:
Update on "Refactor ForeachUtils.h"


--------
- Refactored `can_use_fast_route` logic in ForeachUtils.h. 
- Fixed related bugs in test_foreach.py

Differential Revision: [D26103904](https://our.internmc.facebook.com/intern/diff/D26103904)

[ghstack-poisoned]
